src/psyclone/tests/psyir/transformations/inline_trans_test.py

```diff
@@ -1203,52 +1118,338 @@ def test_validate_named_arg(fortran_reader):
+    '''Test that None is returned when the required Routine is not found
+    in the Container associated with the supplied container symbol, as
+    it does not exist.
+
+    '''
+    code = (
+        f"module inline_mod\n"
+        f"end module inline_mod\n"
+        f"{CALL_IN_SUB_USE}")
+    psyir = fortran_reader.psyir_from_source(code)
+    call_node = psyir.walk(Call)[0]
+    inline_trans = InlineTrans()
+    result = inline_trans._find_routine_in_container(
+            call_node, call_node.routine.interface.container_symbol)
+    assert result is None
+
+
+def test_find_routine_in_container_recurse_named(fortran_reader):
+    '''Test that when a container does not contain the required routine,
+    any imported containers within this container are also
+    searched. In this case the test is for a container within the
+    original container that explicitly names the routine. The PSyIR of
+    the routine is returned when it is found in the second container.
+
+    '''
+    code = (
+        f"{CALL_IN_SUB_USE}"
+        f"module inline_mod\n"
+        f"use inline_mod2, only : sub\n"
+        f"end module inline_mod\n"
+        f"module inline_mod2\n"
+        f"contains\n"
+        f"{SUB}\n"
+        f"end module inline_mod2\n")
+    psyir = fortran_reader.psyir_from_source(code)
+    call_node = psyir.walk(Call)[0]
+    inline_trans = InlineTrans()
+    result = inline_trans._find_routine_in_container(
+            call_node, call_node.routine.interface.container_symbol)
+    assert isinstance(result, Routine)
+    assert result.name == "sub"
+
+
+def test_find_routine_in_container_recurse_wildcard(fortran_reader):
+    '''Test that when a container does not contain the required routine,
+    any imported containers within this container are also
+    searched. In this case the test is for a wildcard container within
+    the original container. The PSyIR of the routine is returned when
+    it is found in the second container.
+
+    '''
+    code = (
+        f"{CALL_IN_SUB_USE}"
+        f"module inline_mod\n"
+        f"use inline_mod2\n"
+        f"end module inline_mod\n"
+        f"module inline_mod2\n"
+        f"contains\n"
+        f"{SUB}\n"
+        f"end module inline_mod2\n")
+    psyir = fortran_reader.psyir_from_source(code)
+    call_node = psyir.walk(Call)[0]
+    inline_trans = InlineTrans()
+    result = inline_trans._find_routine_in_container(
+            call_node, call_node.routine.interface.container_symbol)
+    assert isinstance(result, Routine)
+    assert result.name == "sub"
+
+
+def test_find_routine_in_container_private_routine_not_found(fortran_reader):
+    '''Test that None is returned when the required Routine is not found
+    in the Container associated with the supplied container symbol, as
+    it is private.
+
+    '''
+    private_sub_in_module = SUB_IN_MODULE.replace(
+        "contains\n", "  private :: sub\ncontains\n")
+    code = f"{private_sub_in_module}{CALL_IN_SUB_USE}"
```

Comment:
Strictly, I don't think Fortran would permit a `use xxx, only: sub` if `sub` is private.

Reply:
I just checked this and you are correct. It should therefore be invalid PSyIR. However there are no checks for this at the moment when PSyIR is created or modified. I've added words to that effect in the test's docstring. Is this OK, or do you want me to remove the test? Separately, do you want me to raise an issue for this?

Reply:
We should only spot this problem when attempting to find the definitions of symbols imported from a module. It could be that the associated method already does this? Perhaps @sergisiso knows?

Reply:
In fact, L1192 of symbol_table.py means we do explicitly ignore any private symbols when resolving imports. I consider that to be correct.